zc_events/tests.py

```diff
@@ -0,0 +1,101 @@
+from __future__ import division
+
+import uuid
+import math
+
+from django.conf import settings
+
+from zc_events.utils import event_payload
+
+
+class GlobalIndexRebuildTestMixin(object):
+    """This is a base test case for all index rebuilding events.
+    See in legacy (zerocater.tasks.emit_event_tasks.index_data_events.py module) how to use it.
+    """
+
+    index_rebuild_event_task = None
+    resource_index_rebuild_task = None
+    model = None
+    attributes = []
+    resource_type = None
+    event_name = None
+    serializer = None
+
+    # Optional
+    queryset = None
+    objects_count = 10
+    default_batch_size = 50
+    custom_batch_size = 5
+
+    def setUp(self):
+        super(GlobalIndexRebuildTestMixin, self).setUp()
+        self.create_test_data()
+        self._queryset = self.queryset if self.queryset else self.model.objects.all()
+
+    def create_test_data(self):
+        raise NotImplementedError("Override this method to create test data")
+
+    def test_emitting_event_with_default_batch_size__pass(self, mock_save_string_contents_to_s3,
+                                                          mock_emit_microservice_event):
+        self.resource_index_rebuild_task()
+
+        events_count = int(math.ceil(self.objects_count / self.default_batch_size))
+        self.assertEqual(mock_save_string_contents_to_s3.call_count, events_count)
+        self.assertEqual(mock_emit_microservice_event.call_count, events_count)
+
+    def test_emitting_event_with_custom_batch_size__pass(self, mock_save_string_contents_to_s3,
+                                                         mock_emit_microservice_event):
+        self.resource_index_rebuild_task(batch_size=self.custom_batch_size)
+
+        events_count = int(math.ceil(self.objects_count / self.custom_batch_size))
+        self.assertEqual(mock_save_string_contents_to_s3.call_count, events_count)
+        self.assertEqual(mock_emit_microservice_event.call_count, events_count)
+
+    def test_emitting_event_with_correct_attributes__pass(self, mock_save_string_contents_to_s3,
+                                                          mock_emit_microservice_event):
+        total_events = int(math.ceil(self.objects_count / self.default_batch_size))
+        s3_keys = [str(uuid.uuid4()) for i in xrange(total_events)]
+        mock_save_string_contents_to_s3.side_effect = s3_keys
+
+        self.resource_index_rebuild_task()
+
+        events_count = 0
+        while events_count < total_events:
```

Comment:
Referring to a previous conversation, I don't think this is the best way of testing these methods. Copying the exact implementation of the method is not ideal. I think we can achieve a good level of testing without copying the logic line by line. For instance, if we go and change the logic in the method (referring to my comments above on the method itself), even tho it will still index everything properly this test will fail.